bip3: Switch to SPDX identifiers (#1892)
bitcoin/bips: bip-0003.md

documentation](https://spdx.dev/ids/) and the [SPDX License List](https://spdx.org/licenses/) for further details.

It is also possible to specify that source code is licensed differently by including the optional License-Code header
after the License header. Again, the licensing terms must be specified using an SPDX License Expression.

Each auxiliary source code file or source directory should specify the license under which it is made available as is common in
software (e.g., with a [`SPDX-License-Identifier: <SPDX License Expression>` comment](https://spdx.dev/ids/),
a license header, or a LICENSE/COPYING file). It is recommended to make any test vectors available
under CC0-1.0 or FSFAP in addition to any other licenses to allow anyone to copy test vectors into their
implementations without introducing license hindrances. Licenses listed in the License-Code header apply to all source directories,
source code files, and test vectors provided with the BIP except those where a LICENSE file in a directory
or the file header states otherwise.

For example, a preamble specifying the optional License-Code header might look like:

    License:         CC0-1.0
    License-Code:    MIT

> In this case, the source code in the BIP is not available under Creative Commons CC0 1.0 Universal, but is only available under the MIT

Reviewer: In commit 0cc4d26e675392b49, nit, per https://spdx.org/licenses/CC0-1.0.html

```suggestion
In this case, the source code in the BIP is not available under Creative Commons Zero v1.0 Universal, but is only available under the MIT
```

License.

It is recommended that source code included in a BIP (whether within the text or in auxiliary files) be licensed under the same license terms as the project it
is proposed to modify, if any. For example, changes intended for Bitcoin Core would ideally be licensed (also) under the MIT
License.

In all cases, details of the licensing terms must be provided in the Copyright section of the BIP.

#### Acceptable Licenses[^licenses]

Each new BIP must be made available under at least one acceptable license as listed below. BIPs are not required to be
*exclusively* licensed under approved terms, and may also be licensed under other licenses *in addition to* at least one
acceptable license.

In other words, a new BIP must specify an SPDX License Expression that is either "L" or equivalent to "L OR E" for some
acceptable license L from the following list and another SPDX License Expression E.

* BSD-2-Clause: [BSD 2-Clause License](https://opensource.org/licenses/BSD-2-Clause)
* BSD-3-Clause: [BSD 3-Clause License](https://opensource.org/licenses/BSD-3-Clause)
* CC0-1.0: [Creative Commons CC0 1.0 Universal](https://creativecommons.org/publicdomain/zero/1.0/)

> * FSFAP: [FSF All Permissive License](https://www.gnu.org/prep/maintain/html_node/License-Notices-for-Other-Files.html)

Reviewer: In commit e5748c99971982634, maybe update the link as well, to https://spdx.org/licenses/FSFAP.html

* CC-BY-4.0: [Creative Commons Attribution 4.0 International](https://creativecommons.org/licenses/by/4.0/)
* MIT: [Expat/MIT/X11 License](https://opensource.org/licenses/MIT)
* Apache-2.0: [Apache License 2.0](https://www.apache.org/licenses/LICENSE-2.0)
* BSL-1.0: [Boost Software License 1.0](https://www.boost.org/LICENSE_1_0.txt)

#### Not Acceptable Licenses

All licenses not explicitly included in the above lists are not acceptable terms for a Bitcoin Improvement Proposal.

> However, BIPs predating this proposal were accepted under other terms, and should use one the following identifiers.

Reviewer: In commit 0cc4d26e675392b49b32, missing word

```suggestion
However, BIPs predating this proposal were accepted under other terms, and should use one of the following identifiers.
```

* LicenseRef-PD: Placed into the public domain

> * OPUBL-1.0: [Open Publication License 1.0](https://opencontent.org/openpub/)

Reviewer: In commit 85a248f68eb9, maybe update the link https://spdx.org/licenses/OPUBL-1.0.html

### Specification

Each new BIP must specify in two ways under which license terms it is made available. First, it must specify an [SPDX
License Expression](https://spdx.dev/ids/) in the License field in the preamble. Second, it must include a matching
Copyright section, possibly providing further details on licensing.

For example, a preamble might include the following License header:

    License: CC0-1.0 OR MIT

> In this case, the BIP (including all auxiliary files) is made available under the terms of both Creative Commons CC0 1.0
> Universal as well as the MIT License, and anyone may modify and redistribute it provided they comply with the terms of
> *either* license, at their option. In other words, the license list is an "OR choice", not an "AND also" requirement.
> See the [SPDX documentation](https://spdx.dev/ids/) and the [SPDX License List](https://spdx.org/licenses/) for further
> details.

Reviewer 2: I wholly support all of the improvements of the content, but if you end up touching this again, it would be preferable if diffs were minimized by leaving line-breaks in the same places even when lines are shortened or lengthened by edits affecting a single line as that makes it easier for reviewers to identify what exactly is changing.

```suggestion
In this case, the BIP (including all auxiliary files) is made available under the terms of both Creative Commons CC0 1.0 Universal as well as the
MIT License, and anyone may modify and redistribute it provided they comply with the terms of
*either* license, at their option. In other words, the license list is an "OR choice", not an "AND also" requirement.
See the [SPDX documentation](https://spdx.dev/ids/) and the [SPDX License List](https://spdx.org/licenses/) for further
details.
```

It is also possible to specify that source code is licensed differently by including the optional License-Code header
after the License header. Again, the licensing terms must be specified using an SPDX License Expression.

> Each auxiliary source code file or source directory should specify the license under which it is made available as is
> common in software (e.g., with a [`SPDX-License-Identifier: <SPDX License Expression>` comment](https://spdx.dev/ids/),
> a license header, or a LICENSE/COPYING file). It is recommended to make any test vectors available under CC0-1.0 or
> FSFAP in addition to any other licenses to allow anyone to copy test vectors into their implementations without
> introducing license hindrances. Licenses listed in the License-Code header apply to all source directories, source code
> files, and test vectors provided with the BIP except those where a LICENSE file in a directory or the file header states
> otherwise.

Reviewer 2: Good changes, but in the future, please don’t reformat text passages to adhere to a specific line length. Instead leave the line breaks in the same places (especially the line breaks of otherwise unchanged lines). Due to the shifted line breaks, this edit that adds and changes only a few words (+auxiliary, +[`SPDX-License-Identifier: <SPDX License Expression>` comment](https://spdx.dev/ids/), GNU-All-Permissive ↦ FSFAP) appears in the diff viewer as if the whole paragraph had been rewritten.

```suggestion
Each auxiliary source code file or source directory should specify the license under which it is made available as is common in
software (e.g., with a [`SPDX-License-Identifier: <SPDX License Expression>` comment](https://spdx.dev/ids/), a license header, or a LICENSE/COPYING file). It is recommended to make any test vectors available
under CC0-1.0 or FSFAP in addition to any other licenses to allow anyone to copy test vectors into their
implementations without introducing license hindrances. Licenses listed in the License-Code header apply to all source
directories, source code files, and test vectors provided with the BIP except those where a LICENSE file in a directory
or the file header states otherwise.
```

For example, a preamble specifying the optional License-Code header might look like:

    License:         CC0-1.0
    License-Code:    MIT

In this case, the source code in the BIP is not available under Creative Commons CC0 1.0 Universal, but is only
available under the MIT License.

> It is recommended that source code included in a BIP (whether within the text or in auxiliary files) be licensed under
> the same license terms as the project it is proposed to modify, if any. For example, changes intended for Bitcoin Core
> would ideally be licensed (also) under the MIT License.

Reviewer 2: Nit, should you end up editing further:

```suggestion
It is recommended that source code included in a BIP (whether within the text or in auxiliary files) be licensed under the same license terms as the project it
is proposed to modify, if any. For example, changes intended for Bitcoin Core would ideally be licensed (also) under the MIT
License.
```
